File: frontend/src/services/artistService.js
const API_BASE_URL = "http://localhost:8080/artists";

export async function getRandomArtists() {
  const response = await fetch(`${API_BASE_URL}/random`);
  if (!response.ok) {
    throw new Error("Failed to fetch artists");
  }
  return response.json();
}

export async function searchArtists(query) {
  const response = await fetch(
    `${API_BASE_URL}/search?q=${encodeURIComponent(query)}`
  );
  if (!response.ok) {
    throw new Error("Failed to search artists");
  }
  return response.json();
}

export async function getArtistById(artistId) {
  const response = await fetch(`${API_BASE_URL}/${encodeURIComponent(artistId)}`);
  if (!response.ok) {
    throw new Error("Failed to fetch artist");
  }
  return response.json();
}

export async function getAlbumsForArtist(artistId) {
  const response = await fetch(
    `${API_BASE_URL}/${encodeURIComponent(artistId)}/albums`
  );
  if (!response.ok) {
    throw new Error("Failed to fetch artist albums");
  }
  return response.json();
}

File: backend/src/main/java/com/example/soundwave/service/ArtistService.java
package com.example.soundwave.service;

import com.example.soundwave.dto.musicbrainz.ArtistDto;
import java.util.List;

/**
 * Service interface for artist-related business logic
 *
 * This layer abstracts where artist data comes from
 * (MusicBrainz, database, cache, etc.)
 */

public interface ArtistService {

    /**
     * Returns a list of random artists.
     *
     * @return list of ArtistDto
     */

    List<ArtistDto> getRandomArtists();


    /**
     * Search for artists by name.
     *
     * @param query search keyword
     * @return list of matching artists
     */

    List<ArtistDto> searchArtists(String query);


    /**
     * Fetch a single artist by ID.
     *
     * @param artistId MusicBrainz artist ID
     * @return ArtistDto
     */

    ArtistDto getArtistById(String artistId);
}